instrumentation/spring/spring-jms/spring-jms-2.0/javaagent/src/test/java/SpringListenerTest.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import static io.opentelemetry.api.trace.SpanKind.CONSUMER;
+import static io.opentelemetry.api.trace.SpanKind.PRODUCER;
+import static io.opentelemetry.instrumentation.testing.util.TelemetryDataUtil.orderByRootSpanKind;
+import static io.opentelemetry.sdk.testing.assertj.OpenTelemetryAssertions.equalTo;
+import static io.opentelemetry.sdk.testing.assertj.OpenTelemetryAssertions.satisfies;
+
+import io.opentelemetry.instrumentation.testing.junit.AgentInstrumentationExtension;
+import io.opentelemetry.instrumentation.testing.junit.InstrumentationExtension;
+import io.opentelemetry.sdk.trace.data.LinkData;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import io.opentelemetry.semconv.incubating.MessagingIncubatingAttributes;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.jms.ConnectionFactory;
+import listener.AbstractConfig;
+import listener.AnnotatedListenerConfig;
+import listener.ManualListenerConfig;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+import org.springframework.jms.core.JmsTemplate;
+
+public class SpringListenerTest {
+
+  @RegisterExtension
+  private static final InstrumentationExtension testing = AgentInstrumentationExtension.create();
+
+  @ParameterizedTest
+  @ValueSource(classes = {AnnotatedListenerConfig.class, ManualListenerConfig.class})
+  void receivingMessageInSpringListenerGeneratesSpans(Class<AbstractConfig> config) {
+    AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext(config);
+    ConnectionFactory factory = context.getBean(ConnectionFactory.class);
+    JmsTemplate template = new JmsTemplate(factory);
+
+    template.convertAndSend("SpringListenerJms2", "a message");
+
+    AtomicReference<SpanData> producerSpan = new AtomicReference<>();
+    testing.waitAndAssertSortedTraces(
+        orderByRootSpanKind(PRODUCER, CONSUMER),
+        trace -> {
```

Comment:
Here and other places in this PR, you can extract a public method, and then simplify the code together with the following verification logic. Can refer to https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/1823f147eb0c18ac6259827f16d7808a7dc6774e/instrumentation/xxl-job/xxl-job-common/testing/src/main/java/io/opentelemetry/instrumentation/xxljob/AbstractXxlJobTest.java#L165

Reply:
I think that in the original groovy test the assertions were copied from some other test file. If needed you can move the assertions for produce/receive/process spans to helper methods and share these assertions between tests (if that makes sense), see https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/spring/spring-kafka-2.7/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/spring/kafka/v2_7/SpringKafkaTest.java for ideas.